src/main/java/liqp/Template.java

```diff
@@ -689,4 +698,67 @@ private void putStringKey(boolean convertValueToMap, String key, Object value, M
             map.put(key, value);
         }
     }
+
+    private static List<String> tokenizeString(String input) {
+        List<String> tokens = new ArrayList<>();
+
+        Pattern pattern = Pattern.compile("\\{\\{.*?\\}\\}|\\s+|\\S+");
+        Matcher matcher = pattern.matcher(input);
+        while (matcher.find()) {
+            tokens.add(matcher.group());
+        }
+        return tokens;
+    }
+
+    private static StringBuilder tokenize(String input, Map<String, Object> variables) {
+        StringBuilder result = new StringBuilder();
+
+        Pattern pattern = Pattern.compile("\\{\\{[^{}]*\\}\\}");
+        Matcher matcher = pattern.matcher(input);
+
+        int lastMatchEnd = 0;
+
+        // Iterate through matches and modify them
+        while (matcher.find()) {
+            try {
+                result.append(input, lastMatchEnd, matcher.start());
+                String placeholder = matcher.group();
+
+                Template template = parse(placeholder);
+                String currentResultString = template.render(variables);
+
+                result.append(currentResultString);
+                lastMatchEnd = matcher.end();
+            }
+            catch (Exception exception) {
+                String emptyString = new String();
+                result.replace(matcher.start(), matcher.end(), emptyString);
+                System.err.println("Exception occurred while searching for parsing placeholder strings: " + exception.getMessage());
+            }
+        }
+        result.append(input.substring(lastMatchEnd));
+        return result;
+    }
+
+    private LNode parseRootVisit(ParseTree root, NodeVisitor nodeVisitor) {
+        BlockNode node = new BlockNode();
+
+        if (!Objects.isNull(root) && root.getChildCount() > 0) {
+            ParseTree children = root.getChild(0);
+            int childrenCount = children.getChildCount();
+
+            for (int index = 0; index < childrenCount; index++) {
```

Comment:
can we try using streams here

Reply:
yes for doesn't look good but since it's a ParseTree, stream will not be possible on the object directly.

in the existing liqp library, they are using it as:

public BlockNode visitChildren(RuleNode node) {
        BlockNode result = defaultResult();
        int n = node.getChildCount();
        for (int i = 0; i < n; i++) {
            if (!shouldVisitNextChild(node, result)) {
                break;
            }

            ParseTree c = node.getChild(i);
            try {
                BlockNode childResult = c.accept(this);
                result = aggregateResult(result, childResult);
            } catch (Exception e) {
                // Log the exception or handle it in any other appropriate way
                // You can also return a default value instead of logging/handling the exception
                System.err.println("Exception occurred while visiting child node: " + e.getMessage());
            }
        }
        return result;
    }

Reply:
we can except whatever is there, Lets standarize from new changes so we don't get in more dept pls. See where and all we can use stream and act accordingly at least from our code.

Reply:
actually the ParseTree implements  org.antlr.v4.runtime.tree.SyntaxTree
which doesn't have Iterable and it doesn't provide direct access to stream. 

that's why existing code uses a for loop instead.

ParseTree:
public interface ParseTree extends SyntaxTree {
ParseTree getParent();
ParseTree getChild(int i);
String getText();
String toStringTree(Parser parser);
..
}